docs/man/man1/hal_joystick.1
.\" Copyright (c) 2006 John Kasunich
.\"                (jmkasunich AT users DOT sourceforge DOT net)
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111,
.\" USA.
.\"
.\"
.\"
.TH HAL_JOYSTICK "1"  "2006-03-13" "EMC Documentation" "HAL User's Manual"
.SH NAME
hal_joystick \- \fB(DEPRECATED)\fR control HAL pins with a joystick
.SH SYNOPSIS
\fBhal_joystick\fR [\fB-d\fR \fIdevice\fR] [\fB-p\fR \fIprefix\fR]
.SH DESCRIPTION
\fBhal_joystick\fR is \fBdeprecated\fR.  Use \fBhal_input(1)\fR instead.
\fBhal_joystick\fR will be removed from a future version of emc.

\fBhal_joystick\fR allows a joystick to generate HAL (Hardware Abstraction
Layer) signals.  Although not a hard realtime component, it is quite 
responsive under moderate system load.  It provides analog (float) HAL
pins for each joystick axis, and digital (bit) pins for each joystick
button or trigger.
.SH OPTIONS
.TP
\fB-d\fR \fIdevice\fR
use \fIdevice\fR as the joystick device (default is /dev/input/js0).
.TP
\fB-p\fR \fIprefix\fR
use \fIprefix\fR for the HAL pin names (default is "joystick.0").
.SH USAGE
\fBhal_joystick\fR runs forever until interrupted with SIGINT or SIGTERM.
Normally it would be invoked as \fBhal_joystick &\fR to run in the background.

For each joystick axis, it exports a HAL float pin called "<prefix>.axis.<N>" where N is an integer, starting at zero.  The value of the pin will range from
-1.0 to +1.0 as the axis is moved thru its range of motion.

For each joystick button, it exports a HAL bit pin called "<prefix>.button.<M>"
where M is also an integer starting at zero.

The mapping of axis and buttons to N and M are joystick dependent, as is the
direction of motion that results in positive values of the axis pin.
\fBhal_joystick\fR uses the numbering and direction that is reported by the
Linux joystick driver.  For modern USB or other digital joysticks, the Linux
driver figures out the number of axis and buttons automatically.  For older 
analog joysticks, the driver may need configured by the user.  See Linux
documentation for more details.  Once the Linux driver is properly configured,
the HAL driver will configure itself to match automatically.
.SH "SEE ALSO"
\fBhal_input(1)\fR
.SH BUGS
hal_joystick is incompatible with the way that halcmd waits for components to
be ready.  This leads to race conditions when connecting signals to
hal_joystick's pins.

Perhaps the analog axes should have a "scale" parameter that could be used
to scale the -1.0 to +1.0 range to whatever the user needs. It would also
allow the direction of an axis to be reversed by using a negative scale.
This can already be done using a HAL scale block, but a built-in scale 
parameter would be more convenient.
.SH AUTHOR
Written by John Kasunich, as part of the Enhanced Machine
Controller (EMC) project.
.SH REPORTING BUGS
Report bugs to jmkasunich AT users DOT sourceforge DOT net
.SH COPYRIGHT
Copyright \(co 2003 John Kasunich.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


docs/NEWS
# some of EMC2 History & News
#############################

============ emc2.1.0 =========================================================

2007-01.13 - backported AXIS debug level dialog. Users can now chose
             verbosity directly from AXIS

2007.01.13 - backported HAL driver & sample config for the pluto FPGA 
             board

2007.01.13 - backported AXIS home-all and pyvcp support

2007.01.13 - backported 'unobtainable step rate' error message. when the
	     user requests a velocity which is too high (improper config)
	     this gets reported as a warning during startup.

2006.12.27 - backported streamer (component for data passing from realtime 
             to userspace)

2006.11.17 - removed HAL types s8, u8, s16, and u16.  HAL now uses float
	     for analog values, bit for booleans, and s32 or u32 for
	     integer values.  The only users who will be affected are
	     those with Universal Stepper Controller cards.  See item 9 at
	     http://wiki.linuxcnc.org/cgi-bin/emcinfo.pl?UpdatingConfigurationsForDevelopmentVersions

2006.11.15 - merged classicladder 0.7.100

2006.11.04 - added a feed-hold HAL pin.  Added M codes to enable/disable
	     feed override, spindle override, feed hold, and adaptive feed
	     individually.  Feed override is M50, spindle override is M51,
	     adaptive feed is M52, and feed hold is M53.  G50/51 are no
	     longer used to enable/disable adaptive feed, use M52 instead.

2006.10.24 - added support for block delete as defined by Section 2.2.2
             in the RS274NGC_3 specs.

2006.10.24 - added support for optional program stop as defined by Section 
             2.2.3 in the RS274NGC_3 specs.

2006.10.14 - manpages have been created for most rtapi and hal calls

2006.10.01 - added a userspace simulator which includes rtapi threads.
	     Use "--enable-simulator" to enable this mode, which cannot
	     be used to control real hardware

2006.09.27 - added support to optionally save joint positions between
             runs ([TRAJ] POSITION_FILE = position.txt)

2006.09.26 - added support to specify UNITS in literal in the ini. 
             (e.g. [TRAJ] LINEAR_UNITS = mm )

2006.09.09 - fixed bug #1384883, optional stop was not optional.

2006.08.09 - AXIS is now a part of EMC2, not a separate source download.

2006.08.07 - Added a spindle-speed override. This allows to override the spindle
	     speed during a program run. The MIN_SPINDLE_OVERRIDE and 
	     MAX_SPINDLE_OVERRIDE in the ini assure that speed doesn't get 
	     reduced or increased too much. Both values are in %.

2006.08.06 - Documentation can be built from source with the 'make docs' target

2006.06.11 - added realtime support for jogwheels.  Three HAL pins per axis:
	     one accepts raw counts from the wheel, one sets the distance moved
	     per count, and one enables that axis to move.

2006.05.17 - added 'adaptive feed', a HAL pin that can be used to scale the
	     feedrate in real time.  Simular to feedrate override, but can
	     only vary from 0.0 to 1.0, and reacts much faster.  Intended 
	     for use with EDM and other applications that require closed 
	     loop control of feedrate.

============ emc2.0.0 =========================================================

2006.03.18 - A new pin "count" exists for freqgen.  It is the same as the
             internal symbol "count" shown in the hal introduction.pdf,
             or the result of up OR down if there was an "OR" component.

2006.03.16 - Modified the motion control module to behave more like other
             HAL modules - it no longer adds its functions to threads
             automatically.  All the sample configs have been changed to
             explicitly add the functions, users who have custom configs
             will need to do the same.

2006.03.12 - If a copy of axis 1.3a0 or newer is detected inside src/
             it will be automatically built and installed for
             --enable-run-in-place use

2006.03.10 - minor bugfix in tkemc for jogging

2006.03.10 - added genhexkins, with a minitetra sample config.
             This proves EMC2 works just as good with hexapods as EMC1 did.

2006.03.10 - added VTI support & sample configs

2006.03.07 - lots of fixes to the TP (most were wrong in the old one as well)

2006.03.05 - new Trajectory Planner

2006.02.20 - added a manpage for the runscript.

2006.02.12 - Fixed a bug that was causing the RT code to run very slowly,
             forcing users to set BASE_PERIOD very long.

2006.02.10 - The build process has been replaced, no more recursive $(MAKE)

2006.01.28 - fixed bug #1205237, mode switching sometimes hung

2006.01.08 - M62/M63 digital outputs synchronous to motion
             M64/M65 digital outputs that get set right away

2006.01.06 - work on halconfig.tcl, it is used to display/configure HAL data
             requires the BWidget package, realtime, and HAL to run.

2006.01.04 - fix bug #1387740, a motion that goes nowhere clears feedrate

2006.01.01 - the config picker has i18n support, a German translation exists

2005.12.20 - a new config picker has been written, if you start just the
             runscript this one helps you choose a config to run

2005.12.13 - new runscript, and config organization
             work towards the first release of EMC2

2005.xx.xx - EMC2 now supports tool-changing in complex scenarios

2005.xx.xx - classicladder now a HAL component, a software PLC

2004.xx.xx - HAL (the Hardware Abstraction Layer) is used for
	     addressing hardware drivers, and allows for simple configuration


#developers: this file should hold a history of improvements, bugfixes, etc
             it is useful for keeping track of development


src/hal/user_comps/devices/hal_joystick.c
/********************************************************************
* Description:  hal_joystick.c
*               Joystick driver for HAL
*
* Author: John Kasunich
* License: GPL2
*
* Copyright (c) 2006 All rights reserved.
*
* Last change: 
********************************************************************/

/** 'hal_joystick.c', is a user space driver that allows  HAL pins to
    be controlled by a joystick.  It uses standard Linux  methods to 
    access the joystick, and should work with any joystick that Linux
    recognizes.  It is a user space program, so there is no hard
    realtime guarantee, but under all but the worst loading it is
    quite responsive.
*/

/** Copyright (C) 2006 John Kasunich
                       <jmkasunich AT users DOT sourceforge DOT net>
*/

/** This program is free software; you can redistribute it and/or
    modify it under the terms of version 2 of the GNU General
    Public License as published by the Free Software Foundation.
    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public
    License along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111 USA

    THE AUTHORS OF THIS LIBRARY ACCEPT ABSOLUTELY NO LIABILITY FOR
    ANY HARM OR LOSS RESULTING FROM ITS USE.  IT IS _EXTREMELY_ UNWISE
    TO RELY ON SOFTWARE ALONE FOR SAFETY.  Any machinery capable of 
    harming persons must have provisions for completely removing 
    power from all motors, etc, before persons enter any danger area.
    All machinery must be designed to comply with local and national 
    safety codes, and the authors of this software can not, and do 
    not, take any responsibility for such compliance.

    This code was written as part of the EMC HAL project.  For more
    information, go to www.linuxcnc.org.
*/
#ifndef ULAPI
#error This is a user mode component only!
#endif

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <linux/joystick.h>
#include <errno.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <signal.h>
#include <ctype.h>
#include <string.h>

#include "rtapi.h"		/* RTAPI realtime OS API */
#include "hal.h"		/* HAL public API decls */

#define MAX_AXIS 8
#define MAX_BUTTON 32

typedef struct {
    hal_float_t *axis[MAX_AXIS];
    hal_bit_t *button[MAX_BUTTON];
} hal_js_t;

/* global variables */
int comp_id;		/* HAL component ID */
int done;

/* signal handler */
static void quit(int sig)
{
    done = 1;
}

int main(int argc, char * argv[])
{

    int jsfd, rdlen, n, v, retval;
    struct js_event event;
    int axis_exported[MAX_AXIS];
    int button_exported[MAX_BUTTON];
    hal_js_t *js_data;
    const char *device, *comp_name, *prefix;
    char name[HAL_NAME_LEN + 2];
    struct stat sb;
    int devmajor, devminor, chardev;
    fd_set fds;
    struct timeval tv;

    done = 0;
    /* parse command line */
    device = "/dev/input/js0";
    comp_name = "hal_joystick";
    prefix = "joystick.0";

    n = 1;
    while ( n < argc ) {
	if ( strcmp("-d", argv[n]) == 0 ) {
	    n++;
	    device = argv[n];
	} else if ( strcmp ("-p", argv[n]) == 0 ) {
	    n++;
	    prefix = argv[n];
	    comp_name = argv[n];
	} else {
	    printf ( "Unknown option: '%s'\n", argv[n] );
	    printf ( "Usage: hal_joystick [-d <device>] [-p <prefix>]\n" );
	    return -1;
	} 
    n++;
    }
    /* make sure prefix + suffix (12 chars) fits in HAL_NAME_LEN */
    if ( (strlen(prefix)+12) > HAL_NAME_LEN ) {
	printf ( "ERROR: prefix '%s' is too long, limit is %d chars\n", prefix, HAL_NAME_LEN-12 );
	return -1;
    }

    /* validate device */
    if ( stat (device, &sb) != 0 ) {
	printf ( "ERROR: can't stat() device '%s'\n", device );
	return -1;
    }
    chardev = S_ISCHR(sb.st_mode);
    devmajor = major(sb.st_rdev);
    devminor = minor(sb.st_rdev);
    if ( ! (( chardev ) && ( devmajor == 13 ) && ( devminor <= 31 )) && 
	 ! (( chardev ) && ( devmajor == 15 ) && ( devminor <= 127 )) ) {
	printf ( "ERROR: '%s' is not a joystick\n", device );
	return -1;
    }

    /* STEP 1: open the joystick device file */
    jsfd = open (device, O_RDONLY);
    if ( jsfd < 0 ) {
        printf ( "ERROR: can't open device '%s'\n", device );
        return -1;
    }
    /* STEP 2: initialise the hal component */
    comp_id = hal_init(comp_name);
    if (comp_id < 0) {
	printf( "ERROR: hal_init() failed\n");
	return -1;
    }
    /* Register the routine that catches the SIGINT signal */
    signal(SIGINT, quit);
    /* catch SIGTERM too - the run script uses it to shut things down */
    signal(SIGTERM, quit);
    /* STEP 3: allocate shared memory for joystick data */
    js_data = (hal_js_t *) hal_malloc(sizeof(hal_js_t));
    if (js_data == 0) {
	printf( "ERROR: hal_malloc() failed\n");
	hal_exit(comp_id);
	return -1;
    }
    /* clear the exported flags */
    for ( n = 0 ; n < MAX_AXIS ; n++ ) {
	axis_exported[n] = 0;
    }
    for ( n = 0 ; n < MAX_BUTTON ; n++ ) {
	button_exported[n] = 0;
    }
    /* STEP 4: go into the main loop */
    hal_ready(comp_id);
    while ( !done ) {
	/* use select to wait for joystick action or a timeout */
	FD_ZERO(&fds);
	FD_SET(jsfd, &fds);
	/* timeout is 0.1 second */
	tv.tv_sec = 0;
	tv.tv_usec = 100000;
	/* do the select call */
	retval = select(jsfd+1, &fds, NULL, NULL, &tv );
	/* check for timeout */
	if ( retval == 0 ) {
	    /* timeout, just keep looping (unless 'done' is set */
	    continue;
	}
	/* check for select() errors */
	if ( retval < 0 ) {
	    if ( errno == EINTR ) {
		/* caught a signal, just keep looping */
		continue;
	    }
	    /* a real error */
	    perror ( "ERROR: select() failed: " );
	    break;
	}
	rdlen = read (jsfd, &event, sizeof(struct js_event));
	if ( rdlen != sizeof(struct js_event) ) {
	    printf ( "ERROR: expected %ld bytes, got %d\n", (long)sizeof(struct js_event), rdlen );
	    break;
	}
	/* detect init events, and export a pin for each one */
	if ( event.type & JS_EVENT_INIT ) {
	    if ( event.type & JS_EVENT_AXIS ) {
		n = event.number;
		if ( n < MAX_AXIS ) {
		    if ( !axis_exported[n] ) {
			/* gotta export the HAL pin */
			snprintf(name, HAL_NAME_LEN, "%s.axis.%d", prefix, n);
			retval = hal_pin_float_new(name, HAL_OUT, &(js_data->axis[n]), comp_id);
			if (retval != HAL_SUCCESS) {
			    printf( "ERROR: axis %d export failed with err=%d\n", n, retval);
			    break;
			}
			axis_exported[n] = 1;
		    }
		} else {
		    printf( "WARNING: axis %d exceeds MAX_AXIS (%d)\n", n, MAX_AXIS);
		}
	    } /* if JS_EVENT_AXIS */
	    if ( event.type & JS_EVENT_BUTTON ) {
		n = event.number;
		if ( n < MAX_BUTTON ) {
		    if ( !button_exported[n] ) {
			/* gotta export the HAL pin */
			snprintf(name, HAL_NAME_LEN, "%s.button.%d", prefix, n);
			retval = hal_pin_bit_new(name, HAL_OUT, &(js_data->button[n]), comp_id);
			if (retval != HAL_SUCCESS) {
			    printf( "ERROR: button %d export failed with err=%d\n", n, retval);
			    break;
			}
			button_exported[n] = 1;
		    }
		} else {
		    printf( "WARNING: button %d exceeds MAX_BUTTON (%d)\n", n, MAX_BUTTON);
		}
	    } /* if JS_EVENT_BUTTON */
	} /* if JS_EVENT_INIT */
	/* mask off init bit */
	event.type &= ~JS_EVENT_INIT;
	/* detect change events and write to HAL */
	if ( event.type & JS_EVENT_AXIS ) {
	    n = event.number;
	    v = event.value;
	    if ( n < MAX_AXIS ) {
		*(js_data->axis[n]) = v * ( 1.0 / 32767 );
	    }
	}
	if ( event.type & JS_EVENT_BUTTON ) {
	    n = event.number;
	    v = event.value;
	    if ( n < MAX_BUTTON ) {
		*(js_data->button[n]) = v;
	    }
	}
    } /* main loop */
    /* end of main loop, cleanup and exit */
    close (jsfd);
    hal_exit(comp_id);
    return (0);
}
